drivers/timer/xlnx_psttc_timer.c

```diff
@@ -1,187 +1,186 @@
-#define XTTCPS_CLK_CNTRL_PS_DISABLE	16U  /**< Prescale disable */
-#define XTTCPS_CLK_CNTRL_SRC_MASK	0x00000020U  /**< Clock source */
-#define XTTCPS_CLK_CNTRL_EXT_EDGE_MASK	0x00000040U  /**< External Clock edge */
-
-/* Counter Control Register definitions */
-#define XTTCPS_CNT_CNTRL_DIS_MASK	0x00000001U /**< Disable the counter */
-#define XTTCPS_CNT_CNTRL_INT_MASK	0x00000002U /**< Interval mode */
-#define XTTCPS_CNT_CNTRL_DECR_MASK	0x00000004U /**< Decrement mode */
-#define XTTCPS_CNT_CNTRL_MATCH_MASK	0x00000008U /**< Match mode */
-#define XTTCPS_CNT_CNTRL_RST_MASK	0x00000010U /**< Reset counter */
-#define XTTCPS_CNT_CNTRL_EN_WAVE_MASK	0x00000020U /**< Enable waveform */
-#define XTTCPS_CNT_CNTRL_POL_WAVE_MASK	0x00000040U /**< Waveform polarity */
-#define XTTCPS_CNT_CNTRL_RESET_VALUE	0x00000021U /**< Reset value */
-
-/* Interrupt register masks */
-#define XTTCPS_IXR_INTERVAL_MASK	0x00000001U  /**< Interval Interrupt */
-#define XTTCPS_IXR_MATCH_0_MASK		0x00000002U  /**< Match 1 Interrupt */
-#define XTTCPS_IXR_MATCH_1_MASK		0x00000004U  /**< Match 2 Interrupt */
-#define XTTCPS_IXR_MATCH_2_MASK		0x00000008U  /**< Match 3 Interrupt */
-#define XTTCPS_IXR_CNT_OVR_MASK		0x00000010U  /**< Counter Overflow */
-#define XTTCPS_IXR_ALL_MASK		0x0000001FU  /**< All valid Interrupts */
-
-#define XTTC_MAX_INTERVAL_COUNT 0xFFFFFFFFU /**< Maximum value of interval counter */
-
-static u32_t accumulated_cycles;
-static s32_t _sys_idle_elapsed_ticks = 1;
-
-static int xttc_calculate_interval(u32_t *interval, u8_t *prescaler)
-{
-	u32_t tmpinterval = 0;
-	u8_t tmpprescaler = 0;
-	unsigned int tmpval;
+BUILD_ASSERT_MSG(CYCLES_PER_SEC == CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+		 "Configured system timer frequency does not match the TTC "
+		 "clock frequency");
 
-	tmpval = (u32_t)(TIMER_INPUT_CLKHZ / TIMER_FREQ);
+BUILD_ASSERT_MSG(CYCLES_PER_SEC >= TICKS_PER_SEC,
+		 "Timer clock frequency must be greater than the system tick "
+		 "frequency");
 
-	if (tmpval < (u32_t)65536U) {
-		/* no prescaler is required */
-		tmpinterval = tmpval;
-		tmpprescaler = 0;
-	} else {
-		for (tmpprescaler = 1U; tmpprescaler < 16; tmpprescaler++) {
-			tmpval = (u32_t)(TIMER_INPUT_CLKHZ /
-					 (TIMER_FREQ * (1U << tmpprescaler)));
-			if (tmpval < (u32_t)65536U) {
-				tmpinterval = tmpval;
-				break;
-			}
-		}
-	}
+BUILD_ASSERT_MSG((CYCLES_PER_SEC % TICKS_PER_SEC) == 0,
+		 "Timer clock frequency is not divisible by the system tick "
+		 "frequency");
+
+#ifdef CONFIG_TICKLESS_KERNEL
+static u32_t last_cycles;
+#endif
 
-	if (tmpinterval != 0) {
-		*interval = tmpinterval;
-		*prescaler = tmpprescaler;
-		return 0;
+static u32_t read_count(void)
+{
+	/* Read current counter value */
+	return sys_read32(TIMER_BASE_ADDR + XTTCPS_COUNT_VALUE_OFFSET);
+}
+
+static void update_match(u32_t cycles, u32_t match)
+{
+	/* Ensure that the match value meets the minimum timing requirements */
+	if ((match - cycles) < CYCLES_NEXT_MIN) {
+		match += CYCLES_PER_TICK;
```

Comment:
Note that this assumes CYCLES_PER_TICK > CYCLES_NEXT_MIN.  While that seems true here, someone might want to come along and tune this to reduce the minimum clamp for latency reasons and break it.  I'd suggest an assert here, or alternatively to do the math correctly to advance to the "next tick after the minimum clamp" for all values.

Reply:
I will add an assert here.

Reply:
On second thought, I will just do the proper math here to ensure that the minimum timing requirement is satisfied.